### PR TITLE
[cmake] add PKG_CONFIG_LIBDIR for addon build systems

### PR DESCRIPTION
**Pull request by phunkyfish**
## Description

Copy any Cmake scripts from the addon dependency directory to the source directory.

## Motivation and Context

For addon depends CMakeLists.txt is copied to the source dir of dependency but not any scripts. This is useful if a wrapper script is required or a FindXXX() script.

## How Has This Been Tested?

OSX, on new inputstream addon under developement: https://github.com/phunkyfish/inputstream.ffmpegarchive

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by phunkyfish** [comment]
@Rechi, I'm by no means a cmake expert. Wanted to check if what I was doing is sane and reasonable.

I made the above change when settings up addon depends for ffmpeg in an inputstream addon (dependent on gmp, nettle and gnutls).

**Review by Rechi — CHANGES_REQUESTED** [review]
Find modules are not dependency specific, so it is wrong to have them bound to a specific library.
Also the upstream build system should find it's dependencies on it's own. If that doesn't work it has to be fixed.

I don't know what a wrapper script would do, so I need an example to judge the usage of those.

**Comment by wsnipex** [comment]
I agree with Rechi, I don't see why this is needed

**Comment by phunkyfish** [comment]
As I say I'm not a cmake expert, happy to use an alternative approach. I based the original idea on how FindGnuTls.cmake is copied and used when building ffmpeg in kodi depends.

Here are the examples:

1) For a find script, gnutls finding gmp: https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/Matrix/depends/common/gnutls/FindGmp.cmake. Called from https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/ae344fcd6480a2c5842c22d789c7080dab514d18/depends/common/gnutls/CMakeLists.txt#L7

2) Wrapper script for ffmpeg configure, without it PKG_CONFIG_PATH needs to be set on the environment: https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/Matrix/depends/common/ffmpeg/ConfigureEnvWrapper.cmake. Called from https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/ae344fcd6480a2c5842c22d789c7080dab514d18/depends/common/ffmpeg/CMakeLists.txt#L94-L134

3) Cleanup script after building BZip2: https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/Matrix/depends/common/bzip2/PostInstall.cmake. Called from https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/ae344fcd6480a2c5842c22d789c7080dab514d18/depends/common/bzip2/CMakeLists.txt#L23

**Comment by Rechi** [comment]
1. `find_package(Gmp)` call is useless as the result isn't used.
2. Why workaround this for every dependency instead of fixing it once at the correct place?
3. There is no reason for this cleanup. Having those executables installed doesn't have any effect, but the cleanup script could actually break other dependencies, because of the globing.
If you don't want those files installed disable or patch it away in the upstream build system.

**Comment by phunkyfish** [comment]
On 1 and 3 those makes sense thanks.

Where is the correct place to fix 2? Note that this is only required for FFMpeg. No other dependency requires this workaround.

**Comment by Rechi** [comment]
`PKG_CONFIG_PATH` or actually `PKG_CONFIG_LIBDIR` as environment variable is missing for all dependencies, not only FFmpeg.
It has to be fixed at `tools/depends/target/config-binaddons.site.in` for autotools based buildsystems. For others adding it to `tools/depends/target/Toolchain_binaddons.cmake` could be the fix, but I haven't tested it.

**Comment by phunkyfish** [comment]
Ok, let me try this out. Thanks for the guidance.

**Comment by phunkyfish** [comment]
Even after fixing the PKG_CONFIG issue I will still have a problem here: https://github.com/phunkyfish/inputstream.ffmpegarchive/blob/ae344fcd6480a2c5842c22d789c7080dab514d18/depends/common/ffmpeg/CMakeLists.txt#L87-L90

As I need to check for `gnutls` using `FindGnuTls.cmake`. So is this PR still valid in this case or is there an alternative?

**Comment by phunkyfish** [comment]
Another thing. In `tools/depends/target/config-binaddons.site.in` there are no other variables that I can see that reference a path like `tools/depends/target/binary-addons/macosx10.14_x86_64-target-debug/build/depends/lib/pkgconfig` in the source. 

I'm unsure how to populate the template, I can see `@deps_dir@` but I don't see how to get the source directory preceding that.

**Comment by phunkyfish** [comment]
Latest commit did not work. The correct path the generated but they have no effect on the autotools issue with FFMPEG.

**Comment by phunkyfish** [comment]
So the latest commit setting the values in `Makefile.include.in` works for autotools, however I just modified `Makefile.include` instead of rebuilding. Need to rerun that to check the impact on a depends build. 

Note that only `PKG_CONFIG_PATH` has an effect. `PKG_CONFIG_LIBDIR` has no effect.

Found another issue in that when building GMP libtool does not use the correct path for standard C includes and only picks them up from `/usr/local/include`which is incorrect. I'm not sure where or how to setup the this up to detect the right include path. I think it's similar to the current `PKG_CONFIG_PATH` as it also uses autotools.

**Comment by wsnipex** [comment]
- Makefile.include is not used by addons, so you shouldn't need to touch that.
- xbmc-addons.include is used by addons.
- FFMPEG doesn't use autotools, so while the config-binaddons.site.in change is correct, it won't help for ffmpeg. The easiest way is to set the needed vars in your addons ffmpegs CMakeList.txt

**Comment by phunkyfish** [comment]
Cool, will try `xbmc_addons.include` instead and leave `Makefile.include.in` alone. 

On your last point using `CMakeLists.txt` doesn’t work for ffmpeg unless I use a wrapper script. Nothing set in or passed to an externalProject_add call has any effect as ENV is not passed. So I may be stuck with the workarounds of copying them scripts and using he configure wrapper.

**Comment by Rechi** [comment]
The bundled dependencies aren't a guessing game, you know if gnutls is available or not, as it has to be build via bundled dependencies too.
If you export `PKG_CONFIG_LIBDIR` in `xbmc_addons.include` it should be available at FFmpeg configure time.
Another solution is `CONFIGURE_COMMAND PKG_CONFIG_LIBDIR=<PATH> <CONFIGURE-COMMAND>`.

**Comment by phunkyfish** [comment]
Only adding `PKG_CONFIG_PATH` in `xbmc_addons.include` worked for FFMPEG.

So in each of the 3 locations I set both variables. Copy commit removed and title updated.

**Review by Rechi — COMMENTED** [review]
Setting `PKG_CONFIG_LIBDIR` is enough, as setting `PKG_CONFIG_PATH` in that case to the same value has no effect.
IMO using `@abs_top_srcdir@/target/binary-addons/@deps_dir@` in `Toolchain_binaddons.cmake.in` and `config-binaddons.site.in` is wrong as `tools/depends/xbmc-addons.include` defines the prefix for add-on depends and not `tools/depends/configure.ac`.

**Comment by phunkyfish** [comment]
Ok, so I can limit the change to only `tools/depends/xbmc-addons.include`, however setting only `PKG_CONFIG_LIBDIR` is not sufficient. It will not work unless `PKG_CONFIG_PATH` is also set. At least not on OSX.

**Comment by Rechi** [comment]
Maybe it doesn't work, because you aren't exporting the variable.
I looked at the source code of pkg-config and only setting `PKG_CONFIG_LIBDIR` is enough.

**Comment by phunkyfish** [comment]
Eureka, that was the missing piece.

**Comment by phunkyfish** [comment]
Should be ready now.

**Comment by phunkyfish** [comment]
jenkins build this please

**Comment by phunkyfish** [comment]
Without a mechanism to run anything custom other than a patch how would is be possible to replicate this needed for darwin embedded platforms?

https://github.com/xbmc/xbmc/blob/0bf74b00c9d8d82275707439bb72cd6987932d69/tools/depends/target/ffmpeg/Makefile#L85

**Comment by phunkyfish** [comment]
@Rechi @wsnipex 

Are you ok with this going in? I would like to start working on adding the new binary addon into Jenkins etc but need my first platform to build in CI to get started which requires this change.

File: tools/depends/xbmc-addons.include [hunk 1]
```diff
@@ -1,6 +1,7 @@
 ADDON_DEPS_DIR := $(BUILDDIR)/$(PLATFORM)/build/depends
 TOOLCHAIN_FILE = $(ADDON_DEPS_DIR)/share/Toolchain_binaddons.cmake
 ADDON_PROJECT_DIR = $(BUILDDIR)/../../../../cmake/addons
+export PKG_CONFIG_LIBDIR = $(BUILDDIR)/$(PLATFORM)/build/depends/lib/pkgconfig
```
Comment (wsnipex) [inline tools/depends/xbmc-addons.include]: please use $(ADDON_DEPS_DIR)/lib/pkgconfig, it's more readable
Then it's good to go

Comment (phunkyfish) [reply tools/depends/xbmc-addons.include]: Great, should be ready now.

**Comment by wsnipex** [comment]
Jenkins build this please

**Comment by phunkyfish** [comment]
> Without a mechanism to run anything custom other than a patch how would is be possible to replicate this needed for darwin embedded platforms?
> 
> https://github.com/xbmc/xbmc/blob/0bf74b00c9d8d82275707439bb72cd6987932d69/tools/depends/target/ffmpeg/Makefile#L85
> 
> I guess the only option is to stick with a patch.

A patch won't work for this as the `config.h` file is generated with configure. Any ideas?

**Comment by wsnipex** [comment]
Try defining the BUILD_COMMAND as sed ... && ${CMAKE_MAKE_COMMAND}

**Comment by phunkyfish** [comment]
I went for `external_project_add_step` as the allowed me to do it just for one platform easily:

`
if(CORE_SYSTEM_NAME STREQUAL darwin_embedded)
  externalproject_add_step(ffmpeg
                           RunSedCommand
                           COMMAND echo "Setting HAVE_CLOCK_GETTIME to 0 for darwin_embedded using sed"
                           COMMAND sed -i -- "s/HAVE_CLOCK_GETTIME 1/HAVE_CLOCK_GETTIME 0/g" ${CMAKE_CURRENT_BINARY_DIR}/ffmpeg-prefix/src/ffmpeg-build/config.h
                           DEPENDEES configure)
endif()
`

**Review by wsnipex — APPROVED** [review]

**Pull request merged**